File: java-service/src/main/java/com/nexchat/repository/UserRepository.java
/**
 * NexChat — User Repository
 *
 * Spring Data JPA automatically generates the implementation of this interface
 * at application startup. No SQL or JDBC code required for standard operations.
 *
 * Extends JpaRepository<User, UUID>:
 *   - User: the entity type
 *   - UUID: the primary key type
 *
 * JpaRepository gives us: save(), findById(), findAll(), delete(),
 * existsById(), count(), and more — all out of the box.
 */
package com.nexchat.repository;

import com.nexchat.entity.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface UserRepository extends JpaRepository<User, UUID> {

    /*
     * Derived query: Spring parses "findByEmail" → SELECT u FROM User u WHERE u.email = ?1
     *
     * Returns Optional<User> — never null. Forces the caller to handle the
     * "user not found" case explicitly. This is idiomatic Java 8+ code.
     *
     * Used by: UserDetailsServiceImpl (Spring Security authentication),
     *          AuthService (login flow)
     */
    Optional<User> findByEmail(String email);

    /*
     * Used for username uniqueness check during registration.
     * Exists queries are more efficient than findBy — they stop at the first
     * matching row and return a boolean (no entity hydration needed).
     */
    boolean existsByEmail(String email);

    boolean existsByUsername(String username);

    /*
     * @Query with JPQL: used when derived queries become unreadable.
     * JPQL references entity field names (passwordHash), not column names (password_hash).
     *
     * @Modifying: required for UPDATE/DELETE JPQL queries. Without it, Spring
     * treats the query as a SELECT and throws an exception.
     *
     * @Transactional: UPDATE queries require a transaction. We annotate here
     * (in addition to the service layer) as a safety net.
     *
     * Used by: AuthService when a user changes their password.
     */
    @Modifying
    @Transactional
    @Query("UPDATE User u SET u.passwordHash = :passwordHash, u.updatedAt = :updatedAt WHERE u.id = :userId")
    int updatePasswordHash(
        @Param("userId") UUID userId,
        @Param("passwordHash") String passwordHash,
        @Param("updatedAt") Instant updatedAt
    );

    /*
     * Updates last_seen timestamp. Called by the Go service via REST when
     * a user's WebSocket connection closes.
     *
     * Returns int (rows affected) — good practice for update queries.
     * 0 means user not found; 1 means updated. Anything else is a bug.
     */
    @Modifying
    @Transactional
    @Query("UPDATE User u SET u.lastSeen = :lastSeen WHERE u.id = :userId")
    int updateLastSeen(
        @Param("userId") UUID userId,
        @Param("lastSeen") Instant lastSeen
    );

    /**
     * Returns all users except the current user.
     * Used to populate the contacts sidebar when no search query is provided.
     * Ordered by username for consistent display.
     */
    @Query("SELECT u FROM User u WHERE u.id <> :excludeId ORDER BY u.username ASC")
    List<User> findAllExcept(@Param("excludeId") UUID excludeId);

    /**
     * Case-insensitive username prefix search.
     * Uses LOWER() + LIKE for portability across DBs.
     * PostgreSQL's ILIKE would be faster, but JPQL LOWER+LIKE works on all JPA providers.
     *
     * @param query     The search string (partial username)
     * @param excludeId The current user's ID — exclude from results
     */
    @Query("SELECT u FROM User u WHERE LOWER(u.username) LIKE LOWER(CONCAT('%', :query, '%')) AND u.id <> :excludeId ORDER BY u.username ASC")
    List<User> searchByUsername(
        @Param("query") String query,
        @Param("excludeId") UUID excludeId
    );
}

File: backend/src/repositories/user.repository.ts
/**
 * NexChat — User Repository
 *
 * Replaces: UserRepository.java (Spring Data JPA)
 *
 * Method parity with Java:
 *   findByEmail()        → findByEmail()
 *   existsByEmail()      → existsByEmail()
 *   existsByUsername()   → existsByUsername()
 *   updatePasswordHash() → updatePassword()
 *   updateLastSeen()     → updateLastSeen()
 *   findAllExcept()      → findAllExcept()
 *   searchByUsername()   → searchByUsername()
 *
 * IMPORTANT: passwordHash is NEVER returned by public methods.
 * The only exception is findByEmailWithPassword(), used exclusively
 * by AuthService for login credential verification.
 */
import type { FilterQuery, UpdateQuery } from "mongoose";
import { User, type IUser } from "../models/User.js";
import { NotFoundError } from "../utils/errors.js";

// ── Public (safe) User type — no password hash ────────────────────────────────
export type SafeUser = Omit<IUser, "passwordHash">;

// Fields to exclude when returning users publicly
const EXCLUDE_PASSWORD = { passwordHash: 0 } as const;

// ── Repository ────────────────────────────────────────────────────────────────

/**
 * Find a user by their UUID. Returns null if not found.
 * Excludes passwordHash.
 */
export async function findUserById(id: string): Promise<SafeUser | null> {
  return User.findById(id, EXCLUDE_PASSWORD).lean<SafeUser>();
}

/**
 * Find a user by email — WITHOUT password exclusion.
 * Use ONLY in AuthService for credential verification.
 * Never send this result to the client.
 */
export async function findUserByEmailWithPassword(
  email: string
): Promise<IUser | null> {
  return User.findOne({ email: email.toLowerCase() }).lean<IUser>();
}

/**
 * Find a user by email — safe, no password hash.
 */
export async function findUserByEmail(
  email: string
): Promise<SafeUser | null> {
  return User.findOne({ email: email.toLowerCase() }, EXCLUDE_PASSWORD).lean<SafeUser>();
}

/**
 * Find a user by username — safe, no password hash.
 */
export async function findUserByUsername(
  username: string
): Promise<SafeUser | null> {
  return User.findOne({ username }, EXCLUDE_PASSWORD).lean<SafeUser>();
}

/**
 * Check if an email is already taken (registration check).
 * More efficient than findByEmail — stops at first match, no doc hydration.
 */
export async function existsByEmail(email: string): Promise<boolean> {
  return User.exists({ email: email.toLowerCase() }).then(Boolean);
}

/**
 * Check if a username is already taken (registration check).
 */
export async function existsByUsername(username: string): Promise<boolean> {
  return User.exists({ username }).then(Boolean);
}

/**
 * Create a new user. The passwordHash must already be a BCrypt hash —
 * plain-text passwords MUST NOT be passed here.
 *
 * @returns The created user without passwordHash.
 */
export async function createUser(data: {
  username: string;
  email: string;
  passwordHash: string;
  avatarUrl?: string;
}): Promise<SafeUser> {
  const user = await User.create({
    username: data.username,
    email: data.email.toLowerCase(),
    passwordHash: data.passwordHash,
    avatarUrl: data.avatarUrl,
  });

  // Convert to plain object and strip passwordHash
  const { passwordHash: _omit, ...safe } = user.toObject<IUser>();
  return safe as SafeUser;
}

/**
 * Update a user's password hash (password change flow).
 * Returns the number of documents modified (0 = user not found, 1 = success).
 *
 * Replaces: userRepository.updatePasswordHash() @Modifying JPQL query
 */
export async function updatePassword(
  userId: string,
  newPasswordHash: string
): Promise<number> {
  const result = await User.updateOne(
    { _id: userId },
    { $set: { passwordHash: newPasswordHash } }
  );
  return result.modifiedCount;
}

/**
 * Update a user's lastSeen timestamp.
 * Called by the Go service via /api/internal when a WebSocket disconnects.
 *
 * Replaces: userRepository.updateLastSeen() @Modifying JPQL query
 */
export async function updateLastSeen(
  userId: string,
  lastSeen: Date
): Promise<number> {
  const result = await User.updateOne(
    { _id: userId },
    { $set: { lastSeen } }
  );
  return result.modifiedCount;
}

/**
 * Update a user's profile fields (avatar, status message).
 * Returns the updated user (without passwordHash) or null if not found.
 */
export async function updateProfile(
  userId: string,
  data: { avatarUrl?: string | undefined; statusMessage?: string | undefined }
): Promise<SafeUser | null> {
  const update: UpdateQuery<IUser> = { $set: {} };
  if (data.avatarUrl !== undefined)
    (update.$set as Record<string, unknown>)["avatarUrl"] = data.avatarUrl;
  if (data.statusMessage !== undefined)
    (update.$set as Record<string, unknown>)["statusMessage"] = data.statusMessage;

  return User.findByIdAndUpdate(userId, update, {
    new: true,
    projection: EXCLUDE_PASSWORD,
  }).lean<SafeUser>();
}

/**
 * Find a user by ID, throw NotFoundError if not found.
 */
export async function findUserByIdOrThrow(id: string): Promise<SafeUser> {
  const user = await findUserById(id);
  if (!user) throw new NotFoundError(`User not found: ${id}`);
  return user;
}

/**
 * Return all users except the given userId, sorted by username.
 * Used to populate the contacts/sidebar list.
 *
 * Replaces: userRepository.findAllExcept(excludeId) JPQL query
 */
export async function findAllUsersExcept(
  excludeUserId: string
): Promise<SafeUser[]> {
  return User.find(
    { _id: { $ne: excludeUserId } } as FilterQuery<IUser>,
    EXCLUDE_PASSWORD,
    { sort: { username: 1 } }
  ).lean<SafeUser[]>();
}

/**
 * Case-insensitive partial username search, excluding the current user.
 *
 * Replaces: userRepository.searchByUsername(query, excludeId) JPQL query
 * Uses MongoDB $regex with 'i' flag (case-insensitive).
 */
export async function searchUsersByUsername(
  query: string,
  excludeUserId: string,
  limit: number
): Promise<SafeUser[]> {
  return User.find(
    {
      username: { $regex: query, $options: "i" },
      _id: { $ne: excludeUserId },
    } as FilterQuery<IUser>,
    EXCLUDE_PASSWORD,
    { sort: { username: 1 }, limit }
  ).lean<SafeUser[]>();
}
